File: Python/030__Substring_with_Concatenation_of_All_Words.py
class Solution(object):
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        if not s or not words:
            return []
        freq = {}
        for word in words:
            if word not in freq:
                freq[word] = 0
            freq[word] += 1

        m = len(words[0])
        ans = []
        for i in range(len(s) - m * len(words) + 1):
            if self.is_valid_concatenation(s, i, m, freq):
                ans.append(i)

        return ans

    def is_valid_concatenation(self, s, j, m, freq):
        dic = {}
        n = len(freq)
        while s[j : j + m] in freq and n > 0:
            word = s[j : j + m]
            if word not in dic:
                dic[word] = 0
            dic[word] += 1
            if dic[word] > freq[word]:
                break;
            if dic[word] == freq[word]:
                n -= 1
            j += m
        return n == 0

File: Java/030__Substring_with_Concatenation_of_All_Words.java
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        if (s == null || s.length() == 0 || words == null || words.length == 0) {
            return new ArrayList<>();
        }
        int len = words.length, word_size = words[0].length();
        HashMap<String, Integer> map = new HashMap<>();
        List<Integer> result = new ArrayList<>();

        for (String word : words) {
            map.put(word, map.getOrDefault(word, 0) + 1);
        }
        for (int i = 0; i <= s.length() - (len * word_size); i++) {
            int j = 0;
            HashMap<String, Integer> seen = new HashMap<>();

            while (j < len) {
                int start = i + (j * word_size);
                int end = start + word_size;
                String word = s.substring(start, end);

                if (map.containsKey(word)) {
                    seen.put(word, seen.getOrDefault(word, 0) + 1);
                } else {
                    break;
                }
                if (map.get(word) < seen.get(word)) {
                    break;
                }
                j++;
            }
            if (j == len) {
                result.add(i);
            }
        }
        return result;
    }
}
